### PR TITLE
feat: cancellable actions

### PR DESCRIPTION
**Pull request by amcdnl**
Adds ability to make actions cancellable.

### CODE REVIEW

**Comment by leon** [comment]
Looks good, but we need some unit tests to make sure it works.
what you say @amcdnl?

File: packages/store/src/state-factory.ts [hunk 1]
```diff
@@ -93,7 +98,7 @@ export class StateFactory {
     }
   }
 
-  invokeActions(getState, setState, dispatch, action): any[] {
+  invokeActions(getState, setState, dispatch, actions$, action) {
```
Comment (deebloo) [inline packages/store/src/state-factory.ts]: we should have type annotations

**Comment by deebloo** [comment]
@amcdnl I wrote a test that proves that the  manual way works. when I try to swap it out for cancellable I get:

```
Failed: Cannot read property 'type' of undefined
        TypeError: Cannot read property 'type' of undefined
            at http://localhost:9876/_karma_webpack_/webpack:/packages/store/src/of-action.ts:12:51
            at Array.forEach (<anonymous>)
            at Object.ofAction packages/store/src/of-action.ts:12:16)
            at StateFactory.webpackJsonp../packages/store/src/state-factory.ts.StateFactory.invokeActions packages/store/src/state-factory.ts:120:72)
            at http://localhost:9876/_karma_webpack_/webpack:/packages/store/src/store.ts:108:12
            at http://localhost:9876/_karma_webpack_/webpack:/packages/store/src/compose.ts:22:14
            at Store.webpackJsonp../packages/store/src/store.ts.Store._dispatch packages/store/src/store.ts:117:7)
            at http://localhost:9876/_karma_webpack_/webpack:/packages/store/src/store.ts:34:45
            at Array.map (<anonymous>)
```

**Comment by deebloo** [comment]
update. I think I fixed that error but the test is failing

File: packages/store/src/state-factory.ts [hunk 2]
```diff
@@ -103,15 +108,25 @@ export class StateFactory {
...
+          } else if (result instanceof Promise) {
+            result = fromPromise(result);
+          }
+
+          if (result instanceof Observable) {
+            result = result.pipe(
+              actionMeta.options.cancellable ? takeUntil(actions$.pipe(ofAction(action.type))) : map(r => r)
+            ); // act like a noop
+
             results.push(result);
```
Comment (deebloo) [inline packages/store/src/state-factory.ts]: it looks like results are only being pushed into the array if it is an observable? I am assuming it should always push into the results array?

File: packages/store/tests/dispatch.spec.ts [hunk 3]
```diff
@@ -172,4 +172,53 @@ describe('Dispatch', () => {
       store.dispatch([new Increment()]);
     })
   );
+
```
Comment (deebloo) [inline packages/store/tests/dispatch.spec.ts]: Below is the failing test. the result should be 1

File: packages/store/src/symbols.ts [hunk 4]
```diff
@@ -62,3 +56,10 @@ export interface StoreOptions<T> {
    */
   children?: any[];
 }
+
+export interface ActionOptions {
+  /**
+   * Cancel the previous uncompleted request(s).
+   */
+  cancellable: boolean;
```
Comment (deebloo) [inline packages/store/src/symbols.ts]: what do you think of the "cancelPrevious"? it also avoids to British vs US spelling

**Pull request merged**